### PR TITLE
[SegGPT] Fix seggpt image processor

### LINKED ISSUE #29549: SegGptImageProcessor broken for 2D prompt masks that are not PIL images
### System Info

- `transformers` version: 4.39.0.dev0
- Platform: macOS-14.2.1-arm64-arm-64bit
- Python version: 3.11.7
- Huggingface_hub version: 0.20.2
- Safetensors version: 0.4.1
- Accelerate version: 0.25.0
- Accelerate config: 	not found
- PyTorch version (GPU?): 2.1.2 (False)
- Tensorflow version (GPU?): not installed (NA)
- Flax version (CPU?/GPU?/TPU?): not installed (NA)
- Jax version: not installed
- JaxLib version: not installed
- Using GPU in script?: <fill in>
- Using distributed or parallel set-up in script?: <fill in>

### Who can help?

_No response_

### Information

- [ ] The official example scripts
- [ ] My own modified scripts

### Tasks

- [ ] An officially supported task in the `examples` folder (such as GLUE/SQuAD, ...)
- [ ] My own task or dataset (give details below)

### Reproduction

The bug happens at this [line](https://github.com/huggingface/transformers/blob/main/src/transformers/models/seggpt/image_processing_seggpt.py#L351) because we expect 3-dimensional input, however, prompt masks can be 2D (semantic map) or 3D (in case a palette was already applied). This, however, is just a problem when passing `prompt_masks` as something different than a `PIL.Image`

```python
from transformers import AutoImageProcessor
import numpy as np
model_id = "BAAI/seggpt-vit-large"
image_processor = AutoImageProcessor.from_pretrained(model_id)
img_2d = np.random.rand(18,18)
_ = image_processor(images=None, prompt_masks=img_2d, return_tensors="pt")
```

### Expected behavior

It should work for every valid image input type rather than just for `PIL.Image`

### PR DESCRIPTION
**Pull request by EduardoPach**
# What does this PR do?

Fixes https://github.com/huggingface/transformers/issues/29549 and add appropriate tests to avoid this issue to repeat itself in the future

## Before submitting
- [ ] This PR fixes a typo or improves the docs (you can dismiss the other checks if that's the case).
- [x] Did you read the [contributor guideline](https://github.com/huggingface/transformers/blob/main/CONTRIBUTING.md#create-a-pull-request),
      Pull Request section?
- [ ] Was this discussed/approved via a Github issue or the [forum](https://discuss.huggingface.co/)? Please add a link
      to it if that's the case.
- [ ] Did you make sure to update the documentation with your changes? Here are the
      [documentation guidelines](https://github.com/huggingface/transformers/tree/main/docs), and
      [here are tips on formatting docstrings](https://github.com/huggingface/transformers/tree/main/docs#writing-source-documentation).
- [x] Did you write any new necessary tests?

## Who can review?

Anyone in the community is free to review the PR once the tests have passed. Feel free to tag
members/contributors who may be interested in your PR.

c.c. @amyeroberts

### CODE REVIEW

**Comment by EduardoPach** [comment]
The fix is probably not ideal (the current way with try..except), but just wanted to start the conversation with a working solution

File: src/transformers/models/seggpt/image_processing_seggpt.py [hunk 1]
```diff
@@ -348,7 +348,10 @@ def _preprocess_step(
         size = size if size is not None else self.size
         size_dict = get_size_dict(size)
 
-        images = make_list_of_images(images)
+        try:
+            images = make_list_of_images(images)
+        except ValueError:
+            images = make_list_of_images(images, expected_ndims=2)
```
Comment (amyeroberts) [inline src/transformers/models/seggpt/image_processing_seggpt.py:354]: No - we shouldn't use try / expect here. In fact, I don't think this call is robust if the input image has two dimensions as that's not currently compatible with many of the image transformations. 

It would be better to have two separate methods, `_preprocess_image` and `_preprocess_mask`, each of which call `_preprocess`. `_preprocess_mask` can then use `mask_to_rgb`, before calling `_preprocess` and we can remove this step entirely for when we process images.

**Review by amyeroberts — COMMENTED** [review]
Thanks for working on this and adding tests ❤️ ! 

As you suspected, try / except isn't the way to go, but it should be simple to handle as there's already conditional mask logic in the image processor

File: src/transformers/models/seggpt/image_processing_seggpt.py (hunk 1 above)
Comment (EduardoPach) [reply src/transformers/models/seggpt/image_processing_seggpt.py:354]: `mask_to_rgb` currently expects a list of numpy arrays and this conversion happens on the `_preprocess_step` method with `make_list_of_images` and `to_numpy_array`. The problem is that we are currently accepting/expecting 2D (semantic maps) and 3D (if already converted to RGB with some logic) images for mask prompt. Maybe an easy fix would be to allow the input of the expected dimension for the mask something like `prompt_mask_dimension` and then pass that to `make_list_of_images`, wdyt?

Comment (amyeroberts) [reply src/transformers/models/seggpt/image_processing_seggpt.py:354]: > mask_to_rgb currently expects a list of numpy arrays 

I don't think this is true - `mask_to_rgb` expects a single image

> Maybe an easy fix would be to allow the input of the expected dimension for the mask something like prompt_mask_dimension and then pass that to make_list_of_images, wdyt?

This isn't ideal. It would be better for `mask_to_rgb` accept masks with and without channels and convert as necessary to RGB for both cases.

Comment (EduardoPach) [reply src/transformers/models/seggpt/image_processing_seggpt.py:354]: Ops my bad, `mask_to_rgb` expects a single image you're right. 

But, still, the current problem lies on the `make_list_of_images` even if we split `_preprocess` to `_preprocess_image` and `_preprocess_mask` this problem would still be there and since `make_list_of_images` takes an `ImageInput` as arg which can be either `PIL`, `numpy`, ... and even if we had this logic to parse the input somewhere else it would still be ambiguous if a 3D mask is a batch of zero channel images or a single RGB mask (for inputs other than `PIL`)

Comment (amyeroberts) [reply src/transformers/models/seggpt/image_processing_seggpt.py:354]: OK, I see, so you're saying that the difficulty comes from the face the passed in masks can be `(B, H, W)` or `(B, C, H, W)`? The current approach wouldn't fix this, as it would still make a batch of `(B, H, W)` masks a list of a `[batch]` instead of `[mask_0, mask_1, mask_2, ... ]` 

I would just add a custom `make_list_of_images` for this model, which tasks the images and masks and makes a list of the masks based on the batch size of the input images

Comment (EduardoPach) [reply src/transformers/models/seggpt/image_processing_seggpt.py]: @amyeroberts Yeah, but there's another issue to that in the sense that if the image is 3D it can either be `(B, H, W)` batched semantic map or `(C, H, W)` RGB image. This will only be a problem if we have `ChannelDimension.First` and `B=3`

Therefore, even creating a custom `make_list_of_images` (or `make_list_of_masks` as the problem is in the masks) we'll have ambiguity in the case where `image.ndim==3`.

This is the portion of `make_list_of_images` that raises problems:

```python
if is_valid_image(images):
        if images.ndim == expected_ndims + 1:
            # Batch of images
            images = list(images)
        elif images.ndim == expected_ndims:
            # Single image
            images = [images]
        else:
            raise ValueError(
                f"Invalid image shape. Expected either {expected_ndims + 1} or {expected_ndims} dimensions, but got"
                f" {images.ndim} dimensions."
            )
        return images
```

At this point the images are either `np.ndarray` or a `torch.tensor` in the context of `SegGpt`. 

I've created a new function called `make_list_of_masks` here https://github.com/huggingface/transformers/pull/29550/commits/7ea7a41b02cc8393f1c265b66c2f7091f6f9e0a4 and deal with that, but had to make an assumption that if `images` is 3D then if the last of first dim is 3 it's a single RGB image.

**Comment by EduardoPach** [comment]
c.c. @NielsRogge

**Review by NielsRogge — APPROVED** [review]
Thanks for improving this!

File: tests/models/seggpt/test_image_processing_seggpt.py [hunk 2]
```diff
@@ -229,3 +231,66 @@ def test_pixel_values(self):
             torch.allclose(inputs.prompt_pixel_values[0, :, :3, :3], expected_prompt_pixel_values, atol=1e-4)
         )
         self.assertTrue(torch.allclose(inputs.prompt_masks[0, :, :3, :3], expected_prompt_masks, atol=1e-4))
+
+    def test_prompt_mask_equivalence(self):
```
Comment (amyeroberts) [inline tests/models/seggpt/test_image_processing_seggpt.py:235]: Very nice!

**Review by amyeroberts — COMMENTED** [review]
Very nice - thanks for working on this! 

Only comment is this would still fail for batch_size = 3.

File: tests/models/seggpt/test_image_processing_seggpt.py (hunk 2 above)
Comment (EduardoPach) [reply tests/models/seggpt/test_image_processing_seggpt.py:235]: thanks!

File: src/transformers/models/seggpt/image_processing_seggpt.py [hunk 3]
```diff
@@ -28,21 +28,75 @@
...
+        return [images]
+
+    if is_valid_image(images):
+        if images.ndim == 2:
+            # Singe semantic map
+            images = [images]
+        elif images.ndim == 3:
+            # Either single RGB or batched semantic map
+            # Assuming that if first or last dimension is 3 (Channel first or last) it is RGB
+            images = [images] if (images.shape[0] == 3 or images.shape[-1] == 3) else list(images)
```
[the remark this answers is not in the document]
Comment (EduardoPach) [reply src/transformers/models/seggpt/image_processing_seggpt.py]: Indeed not ideal, therefore I made a new proposal here https://github.com/huggingface/transformers/pull/29550/commits/f372a40a4fce61e85ec0ce61e4a8a63e445b8e6a

Instead of having to deal with that, I added a new optional argument called `segmentation_maps` that follows other image processor's conventions. So, to obtain the `prompt_masks` in the `BatchedFeature` one should specify `prompt_masks` or `segmentation_maps` where `segmentation_maps` should be the 2-dimensional input that we convert to RGB when `num_labels` is specified.

File: docs/source/en/model_doc/seggpt.md [hunk 4]
```diff
@@ -26,7 +26,8 @@ The abstract from the paper is the following:
 
 Tips:
 - One can use [`SegGptImageProcessor`] to prepare image input, prompt and mask to the model.
-- It's highly advisable to pass `num_labels` (not considering background) during preprocessing and postprocessing with [`SegGptImageProcessor`] for your use case.
+- When using binary masks or segmentation maps as prompts pass it to [`SegGptImageProcessor`] `segmentation_maps` argument, if you already converted your prompt mask to RGB then pass it to `prompt_masks` argument.
+- It's highly advisable to pass `num_labels` when using `segmetantion_maps` (not considering background) during preprocessing and postprocessing with [`SegGptImageProcessor`] for your use case.
```
Comment (amyeroberts) [inline docs/source/en/model_doc/seggpt.md:30]: No, let's not do this. Having two different inputs like this is both error prone and an antipattern. The image processor can selectively process based on the `do_xxx` flags e.g. if the input image was already resized, we wouldn't have a different input e.g. `resized_images` we would still pass in `images` and set `do_resize=False`. In this case, it looks like what we want is `do_convert_rgb`

Comment (EduardoPach) [reply docs/source/en/model_doc/seggpt.md:30]: It makes perfect sense! Introduced the `do_convert_rgb` flag and updated the tests accordingly here https://github.com/huggingface/transformers/pull/29550/commits/2185b3e47de9714d6fa5213ddd30d7dc73427a94

File: src/transformers/models/seggpt/image_processing_seggpt.py [hunk 5]
```diff
@@ -461,6 +441,15 @@ def preprocess(
...
                 Image standard deviation to use if `do_normalize` is set to `True`.
+            do_convert_rgb (`bool`, *optional*, defaults to `self.do_convert_rgb`):
+                Whether to convert the prompt mask to RGB format. If `num_labels` is specified, a palette will be built
+                to map the prompt mask from a single channel to a 3 channel RGB. If unset, the prompt mask is duplicated
+                across the channel dimension.
+            num_labels: (`int`, *optional*):
+                Number of classes in the segmentation task (excluding the background). If specified, a palette will be
+                built, assuming that class_idx 0 is the background, to map the prompt mask from a single class_idx
+                channel to a 3 channel RGB. Not specifying this will result in the prompt mask either being passed
+                through as is if it is already in RGB format or being duplicated across the channel dimension.
```
Comment (amyeroberts) [inline src/transformers/models/seggpt/image_processing_seggpt.py:452]: Is the RGB info here still correct?

File: docs/source/en/model_doc/seggpt.md [hunk 6]
```diff
@@ -26,7 +26,8 @@ The abstract from the paper is the following:
 
 Tips:
 - One can use [`SegGptImageProcessor`] to prepare image input, prompt and mask to the model.
-- It's highly advisable to pass `num_labels` (not considering background) during preprocessing and postprocessing with [`SegGptImageProcessor`] for your use case.
+- When using binary masks or segmentation maps as prompts pass it to [`SegGptImageProcessor`] `segmentation_maps` argument, if you already converted your prompt mask to RGB then pass it to `prompt_masks` argument.
```
Comment (amyeroberts) [inline docs/source/en/model_doc/seggpt.md:29]: To be updated

**Review by amyeroberts — COMMENTED** [review]
Looking good - thanks for iterating on this! 

Just a few places left to tidy up

File: src/transformers/models/seggpt/image_processing_seggpt.py [hunk 7]
```diff
@@ -65,29 +67,10 @@ def build_palette(num_labels: int) -> List[Tuple[int, int]]:
...
-    mask: np.ndarray,
-    palette: Optional[List[Tuple[int, int]]] = None,
-    input_data_format: Optional[ChannelDimension] = None,
-    data_format: Optional[ChannelDimension] = None,
+    mask: np.ndarray, palette: Optional[List[Tuple[int, int]]] = None, data_format: Optional[ChannelDimension] = None
 ) -> np.ndarray:
-    if input_data_format is None and mask.ndim > 2:
-        input_data_format = infer_channel_dimension_format(mask)
-
-    data_format = data_format if data_format is not None else input_data_format
```
[the remark this answers is not in the document]
Comment (EduardoPach) [reply src/transformers/models/seggpt/image_processing_seggpt.py:85]: We only get to this point if `do_convert_rgb` is true, thus the mask, in that case, has no channels. If the image is RGB already then we just apply the other transformations

File: src/transformers/models/seggpt/image_processing_seggpt.py [hunk 8]
```diff
@@ -203,21 +188,11 @@ def mask_to_rgb(
                 image is used. Can be one of:
                 - `"channels_first"` or `ChannelDimension.FIRST`: image in (num_channels, height, width) format.
                 - `"channels_last"` or `ChannelDimension.LAST`: image in (height, width, num_channels) format.
-            input_data_format (`ChannelDimension` or `str`, *optional*):
-                The channel dimension format for the input image. If unset, the channel dimension format is inferred
-                from the input image. Can be one of:
-                - `"channels_first"` or `ChannelDimension.FIRST`: image in (num_channels, height, width) format.
-                - `"channels_last"` or `ChannelDimension.LAST`: image in (height, width, num_channels) format.
```
[the remark this answers is not in the document]
Comment (EduardoPach) [reply src/transformers/models/seggpt/image_processing_seggpt.py:210]: This method is specific for `2D` (no channels) masks, thus I see no reason for it to have an `input_data_format` should I add it to keep the same signature?

File: docs/source/en/model_doc/seggpt.md (hunk 6 above)
Comment (EduardoPach) [reply docs/source/en/model_doc/seggpt.md]: https://github.com/huggingface/transformers/pull/29550/commits/017bd48b4e9c51ea8161d757d47a89ed7c7e8ca2

File: src/transformers/models/seggpt/image_processing_seggpt.py (hunk 5 above)
Comment (EduardoPach) [reply src/transformers/models/seggpt/image_processing_seggpt.py]: It was, but updated nonetheless here https://github.com/huggingface/transformers/pull/29550/commits/017bd48b4e9c51ea8161d757d47a89ed7c7e8ca2

File: src/transformers/models/seggpt/image_processing_seggpt.py (hunk 8 above)
[the remark this answers is not in the document]
Comment (amyeroberts) [reply src/transformers/models/seggpt/image_processing_seggpt.py:210]: OK - in this case let's leave as-is

File: src/transformers/models/seggpt/image_processing_seggpt.py (hunk 7 above)
[the remark this answers is not in the document]
Comment (amyeroberts) [reply src/transformers/models/seggpt/image_processing_seggpt.py:85]: OK, let's leave as-is

**Review by amyeroberts — APPROVED** [review]
Thanks for working on this, fixing the code to make the users experience better and adding tests ❤️
